### PR TITLE
Add QuadTransformerDistributeIri

### PR DESCRIPTION
**Pull request by wvdemeer**
A quad transformer that distributes IRIs over multiple destination IRIs. It does so by interpreting the first capture group as a number and using `replacementStrings[number % replacementStrings.length]`

(credit @twalcari)

### CODE REVIEW

File: lib/transform/QuadTransformerDistributeIri.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    super();
+    this.search = new RegExp(searchRegex, 'u');
+    this.replacements = replacementStrings;
+  }
+
+  protected transformTerm(term: RDF.Term): RDF.Term {
+    if (term.termType === 'NamedNode') {
+      const match = this.search.exec(term.value);
+      if (match) {
+        const nr = Number.parseInt(match[1], 10);
```
Comment (rubensworks) [inline lib/transform/QuadTransformerDistributeIri.ts:27]: Here we assume that the regex will only match numbers, but the user _could_ also provide a regex that matches strings.
I'd suggest adding a try-catch around this to check for number parsing failures, and re-throw a new error saying that the user provided an unexpected regex that matches non-numbers.
Additionally, we should also document this number-matching requirement on the regex.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Thanks, looks great!

Could you also add a brief README entry on this with an example and description of how this new transformer can be used?

**Comment by wvdemeer** [comment]
Added info to `README.md`, and added clear errors to `QuadTransformerDistributeIri.transformTerm`.
Let me know if there's anything else that needs to be changed.

File: lib/transform/QuadTransformerDistributeIri.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  protected transformTerm(term: RDF.Term): RDF.Term {
+    if (term.termType === 'NamedNode') {
+      const match = this.search.exec(term.value);
+      if (match) {
+        if (match.length < 2) {
+          throw new Error(`QuadTransformerDistributeIri error: The "searchRegex" did not contain any groups. ` +
+              `QuadTransformerDistributeIri requires at least one group-based replacement, ` +
+              `of which the first group must match a number.`);
```
Comment (rubensworks) [inline lib/transform/QuadTransformerDistributeIri.ts:33]:
```suggestion
          throw new Error(`'searchRegex' did not contain any groups, while QuadTransformerDistributeIri requires at least one group-based replacement, of which the first group must match a number.`);
```

File: lib/transform/QuadTransformerDistributeIri.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    if (term.termType === 'NamedNode') {
+      const match = this.search.exec(term.value);
+      if (match) {
+        if (match.length < 2) {
+          throw new Error(`'searchRegex' did not contain any groups, while QuadTransformerDistributeIri requires at least one group-based replacement, of which the first group must match a number.`);
+        }
+        const nr = Number.parseInt(match[1], 10);
+        if (Number.isNaN(nr)) {
+          throw new Error(`QuadTransformerDistributeIri error: The first capture group in "searchRegex"` +
+              ` must always match a number, but it matched "${match[1]}" instead.`);
```
Comment (rubensworks) [inline lib/transform/QuadTransformerDistributeIri.ts:36]:
```suggestion
          throw new Error(`The first capture group in 'searchRegex' must always match a number, but it matched "${match[1]}" instead.`);
```

**Comment by rubensworks** [comment]
Thanks for the update!

Could you add unit tests for those two error cases? (The CI required 100% test coverage)

**Comment by wvdemeer** [comment]
100% test coverage now.

**Comment by rubensworks** [comment]
Nice, thanks @wvdemeer!

**Pull request merged**